platform/wayland/cog-platform-wl.c

```diff
@@ -757,14 +1014,17 @@ registry_global (void               *data,
         wl_data.subcompositor = wl_registry_bind(registry, name, &wl_subcompositor_interface, version);
     } else if (strcmp(interface, wl_shell_interface.name) == 0) {
         wl_data.shell = wl_registry_bind(registry, name, &wl_shell_interface, version);
+        win_data.shell_context = (shell_context *) (&wayland_data);
     } else if (strcmp(interface, wl_shm_interface.name) == 0) {
         wl_data.shm = wl_registry_bind(registry, name, &wl_shm_interface, version);
     } else if (strcmp(interface, xdg_wm_base_interface.name) == 0) {
         wl_data.xdg_shell = wl_registry_bind(registry, name, &xdg_wm_base_interface, version);
         g_assert(wl_data.xdg_shell);
         xdg_wm_base_add_listener(wl_data.xdg_shell, &xdg_shell_listener, NULL);
+        win_data.shell_context = (shell_context *) (&xdg_data);
     } else if (strcmp(interface, zwp_fullscreen_shell_v1_interface.name) == 0) {
         wl_data.fshell = wl_registry_bind(registry, name, &zwp_fullscreen_shell_v1_interface, version);
+        win_data.shell_context = (shell_context *) (&f_data);
```

Comment:
These look good, but we're still missing a default setting for win_data.shell_context - if none of these clauses that set it are hit, the first call will likely result in a crash.